plugins/rts54hub/fu-rts54hub-rtd21xx-device.c

```diff
@@ -0,0 +1,618 @@
+	return TRUE;
+}
+
+static gboolean
+fu_rts54hub_rtd21xx_device_attach (FuDevice *device, GError **error)
+{
+	FuRts54hubRtd21xxDevice *self = FU_RTS54HUB_RTD21XX_DEVICE (device);
+	FuRts54HubDevice *parent = FU_RTS54HUB_DEVICE (fu_device_get_parent (device));
+	guint8 buf[] = { ISP_CMD_FW_UPDATE_RESET };
+	g_autoptr(FuDeviceLocker) locker = NULL;
+
+	/* open device */
+	locker = fu_device_locker_new (parent, error);
+	if (locker == NULL)
+		return FALSE;
+	if (!fu_rts54hub_rtd21xx_device_i2c_write (self,
+						   UC_FOREGROUND_SLAVE_ADDR,
+						   UC_FOREGROUND_OPCODE,
+						   buf, sizeof(buf),
+						   error)) {
+		g_prefix_error (error, "failed to attach: ");
+		return FALSE;
+	}
+
+	/* success */
+	fu_device_remove_flag (device, FWUPD_DEVICE_FLAG_IS_BOOTLOADER);
+	return TRUE;
+}
+
+static gboolean
+fu_rts54hub_rtd21xx_device_write_firmware (FuDevice *device,
+					   FuFirmware *firmware,
+					   FwupdInstallFlags flags,
+					   GError **error)
+{
+	FuRts54hubRtd21xxDevice *self = FU_RTS54HUB_RTD21XX_DEVICE (device);
+	const guint8 *fwbuf;
+	gsize fwbufsz = 0;
+	guint32 project_addr;
+	guint8 project_id_count;
+	guint8 read_buf[10] = { 0 };
+	guint8 write_buf[ISP_PACKET_SIZE] = {0};
+	g_autoptr(FuDeviceLocker) locker = NULL;
+	g_autoptr(GBytes) fw = NULL;
+	g_autoptr(GPtrArray) chunks = NULL;
+
+	/* open device */
+	locker = fu_device_locker_new (self, error);
+	if (locker == NULL)
+		return FALSE;
+
+	/* simple image */
+	fw = fu_firmware_get_image_default_bytes (firmware, error);
+	if (fw == NULL)
+		return FALSE;
+	fwbuf = g_bytes_get_data (fw, &fwbufsz);
+
+	/* enable ISP high priority */
+	write_buf[0] = ISP_CMD_ENTER_FW_UPDATE;
+	write_buf[1] = 0x01;
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_BUSY);
+	if (!fu_rts54hub_rtd21xx_device_i2c_write (self,
+						   UC_FOREGROUND_SLAVE_ADDR,
+						   UC_FOREGROUND_OPCODE,
+						   write_buf, 2,
+						   error)) {
+		g_prefix_error (error, "failed to enable ISP: ");
+		return FALSE;
+	}
+	if (!fu_rts54hub_rtd21xx_device_read_status (self, NULL, error))
+		return FALSE;
+
+	/* get project ID address */
+	write_buf[0] = ISP_CMD_GET_PROJECT_ID_ADDR;
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_READ);
+	if (!fu_rts54hub_rtd21xx_device_i2c_write (self,
+						   UC_FOREGROUND_SLAVE_ADDR,
+						   UC_FOREGROUND_OPCODE,
+						   write_buf, 1,
+						   error)) {
+		g_prefix_error (error, "failed to get project ID address: ");
+		return FALSE;
+	}
+
+	/* read back 6 bytes data */
+	g_usleep (I2C_DELAY_AFTER_SEND * 40);
+	if (!fu_rts54hub_rtd21xx_device_i2c_read (self,
+						  UC_FOREGROUND_SLAVE_ADDR,
+						  UC_FOREGROUND_STATUS,
+						  read_buf, 6,
+						  error)) {
+		g_prefix_error (error, "failed to read project ID: ");
+		return FALSE;
+	}
+	if (read_buf[0] != ISP_STATUS_IDLE_SUCCESS) {
+		g_set_error (error, G_IO_ERROR, G_IO_ERROR_INVALID_DATA,
+			     "failed project ID with error 0x%02x: ",
+			     read_buf[0]);
+		return FALSE;
+        }
```

Comment:
whitespace

Reply:
Well spotted! Fixed.

Reply:
> Well spotted! Fixed.

Experimenting with `Github pull requests` extension in vscode.  You get the editor right in front of you with the whitespace settings you use and everything during the PR.  Wouldn't have seen this otherwise :)